File: LeetCodeRecord/ACFirst/142-DetectCycle.java
package ACFirst;

class DetectCycle {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode detectCycle(ListNode head) {
        ListNode dummy = new ListNode(-1);
        ListNode slow = dummy;
        ListNode fast = dummy;
        fast.next = head;

        while (true) {
            if (fast == null || fast.next == null) {
                return null;
            }

            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast) {
                break;
            }
        }

        // 这边的 fast 重新定位需要跟一开始的位置同步
        fast = dummy;
        
        while (fast != slow) {
            fast = fast.next;
            slow = slow.next;
        }

        return slow;
    }
}

File: Top200/142-DetectCycle.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	dummy := &ListNode{Val: -1, Next: head}
	slow := dummy
	fast := dummy

	for true {
		if fast == nil || fast.Next == nil {
			return nil
		}
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			break
		}
	}

	if fast == nil || fast.Next == nil {
		return nil
	}

	slow = dummy
	for slow != fast {
		slow = slow.Next
		fast = fast.Next
	}

	return slow
}
